toc/rfc/rfc-draft-ssh-over-ws.md

```diff
@@ -0,0 +1,269 @@
+- SSH Proxy SHOULD provide a separate handler for unwrapping WebSocket connections:
+```
+	sshWsProxy := wsProxy.New(logger, proxySSHServerConfig, metronClient, tlsConfig)
+	wsServer := server.NewServer(logger, sshProxyConfig.Address, websocket.Handler(sshWsProxy), time.Duration(sshProxyConfig.IdleConnectionTimeout))
+```
+Changes in [ssh-proxy/main.go](https://github.com/cloudfoundry/diego-ssh/blob/main/cmd/ssh-proxy/main.go#L65C1-L67C1)
+
+```
+func (p *WsProxy) HandleConnection(wsConn *websocket.Conn) {
+	logger := p.logger.Session("handle-ws-connection")
+	defer wsConn.Close()
+
+    // Here we start a SSH handshake inside the WebSocket tunnel
+	serverConn, serverChannels, serverRequests, err := ssh.NewServerConn(wsConn, p.serverConfig)
+
+(...)
+```
+Changes in [proxy.go](https://github.com/cloudfoundry/diego-ssh/blob/main/proxy/proxy.go#L67)
+
+**Remark:**
+(these changes are not meant as verbatim code; they only illustrate areas that would need to change)
+
+
+- SSH Proxy MUST retain the legacy SSH server on port 2222 for backwards compatibiliy
+- Diego Release SHOULD provide [a spec](https://github.com/cloudfoundry/diego-release/blob/develop/jobs/ssh_proxy/spec#L53) to define a `diego.ssh_proxy.listen_ws_addr` property with a default of `0.0.0.0:443` or `0.0.0.0:9443`
+
+#### CF Deployment
+
+- CF Deployment [deploys SSH Proxy on Scheduler](https://github.com/cloudfoundry/cf-deployment/blob/main/cf-deployment.yml#L1329) but there is no route to it at the moment.
+- CF Deployment SHOULD deploy [route-registrar](https://github.com/cloudfoundry/routing-release/tree/develop/jobs/route_registrar) alongside SSH Proxy and configure it to announce a TLS route at Gorouter:
+```
+  - name: route_registrar
+    properties:
+      nats:
+        tls:
+          enabled: true
+          client_cert: "((nats_client_cert.certificate))"
+          client_key: "((nats_client_cert.private_key))"
+      route_registrar:
+        routes:
+        - name: ssh-ws-proxy
+          port: 443     # or 9443 if preferred
+          tls_port: 443 # or 9443 if preferred
+          registration_interval: 20s
+          server_cert_domain_san: ssh.((system_domain))
+          uris:
+          - ssh.((system_domain))       # e.g. ssh.cf-app.com, announced by CAPI as wss://ssh.cf-app.com
+    release: routing
+```
+
+
+### Changes needed for regular SSH, SCP and SFTP
+
+- Getting the ssh token via `cf ssh-code` remains unchanged as it connects to UAA.
+- Regular ssh, scp and sftp commands SHOULD support a `-o ProxyCommand` option.
+- The program run as `ProxyCommand` MUST read from `stdin` and write to `stdout` for ssh to work.
+- The CF community MAY offer a simple WebSocket proxy for regular ssh as a [homebrew formular](https://github.com/cloudfoundry/homebrew-tap) or similar option.
+- A simple WebSocket proxy MAY look like this:
```

Comment:
```suggestion
- A simple WebSocket proxy could look like this:
```

This is not an optional specification as part of this RFC but rather a supplement / example to guide implementers, as such it shouldn't use `MAY`.